{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The google.cloud.bigquery extension is already loaded. To reload it, use:\n",
      "  %reload_ext google.cloud.bigquery\n"
     ]
    }
   ],
   "source": [
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gnomAD Queries\n",
    "## Type1: Explore a particular genomic region\n",
    "This category include queries that extract information from a region of genome, for example a gene. Becuase gnomAD BigQuery tables utilize [integer range partitioning](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions) they are optimized for this type of queries.\n",
    "\n",
    "The main requirement to use this feature is to limit queries to a particular region by adding these conditions to the `WHERE` clause:\n",
    "\n",
    "`WHERE start_position >= X AND start_position <= Y`\n",
    "\n",
    "Where `[X, Y]` the region of interst. \n",
    "\n",
    "You can find values of `X` and `Y` by refering to an external databses. For example following table sumarizes the start and end positions for 4 genes on chromosome 17 extracted from an external resource:\n",
    "\n",
    "| Gene \t| X \t| Y \t| Source \t|\n",
    "|:-:\t|-\t|-\t|-\t|\n",
    "| BRCA1 \t| 43044295 \t| 43125364 \t| [link](https://ghr.nlm.nih.gov/gene/BRCA1#location) \t|\n",
    "| COL1A1 \t| 50184096 \t| 50201649 \t| [link](https://ghr.nlm.nih.gov/gene/COL1A1#location) \t|\n",
    "| TP53 \t| 31094927 \t| 31377677 \t| [link](https://ghr.nlm.nih.gov/gene/TP53#location) \t|\n",
    "| NF1 \t| 56593699 \t| 56595611 \t| [link](https://ghr.nlm.nih.gov/gene/NF1#location) \t|\n",
    "\n",
    "Alternatively you could use the following query that extract the same infomration directly from gnomAD tables. \n",
    "\n",
    "In the following example we are using `BRCA1` on `chr17` as example, you could enter your gene of interest to modify all the following queries. Make sure for your gene of interest you are querying the right table (chromosome). If your query returns `NaN` this might be becuase you are querying the wrong table.\n",
    "\n",
    "Also you could choose which version of gnomAD dataset you'd like to user for all the following queries:\n",
    " * `v2_1_1_exomes`\n",
    " * `v2_1_1_genomes`\n",
    " * `v3_genomes`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e961a6191f234c02a178eda9ad4ce419",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dropdown(description='chromosome:', index=16, options=('chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7',…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ToDo: Use ipywidgets to get inputs from user\n",
    "# https://ipywidgets.readthedocs.io/en/latest/examples/Using%20Interact.html\n",
    "\n",
    "import ipywidgets as widgets\n",
    "\n",
    "widgets.Dropdown(\n",
    "    options=['v2_1_1_exomes', 'v2_1_1_genomes', 'v3_genomes'],\n",
    "    value='v3_genomes',\n",
    "    description='gnomAD version:',\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "widgets.Dropdown(\n",
    "    options=['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY'],\n",
    "    value='chr17',\n",
    "    description='chromosome:',\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "widgets.Text(\n",
    "    value='BRCA1',\n",
    "    placeholder='gene_symbol',\n",
    "    description='Gene Symbol:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "from ipywidgets import interact\n",
    "\n",
    "def f(x):\n",
    "    return x\n",
    "\n",
    "interact(f, x='BRCA1');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ToDo: Use ipywidgets to get inputs from user\n",
    "chromosome = 'chr17'\n",
    "gene_symbol = 'BRCA1'\n",
    "gnomad_version = 'v3_genomes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from google.cloud import bigquery\n",
    "\n",
    "client = bigquery.Client()\n",
    "\n",
    "def run_query(query):\n",
    "    query_job = client.query(query)\n",
    "    result = query_job.to_dataframe(progress_bar_type='tqdm_notebook')\n",
    "    cost_cents = (query_job.total_bytes_billed / 1024 ** 4) * 500\n",
    "    print('This query costed (¢): {}'.format(cost_cents))\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "63f599c92680447cac9edfa1508b955b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=1.0, style=ProgressStyle(description_wi…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed: 0.0018978118896484375$\n",
      "          X         Y\n",
      "0  41191318  41282493\n"
     ]
    }
   ],
   "source": [
    "query_template = \"\"\"\n",
    "SELECT MIN(start_position) AS X, MAX(end_position) AS Y\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table\n",
    "WHERE EXISTS\n",
    "  (SELECT 1 FROM UNNEST(main_table.alternate_bases) AS alternate_bases\n",
    "   WHERE EXISTS (SELECT 1 from alternate_bases.vep WHERE SYMBOL = '{GENE}'))\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, GENE=gene_symbol)\n",
    "\n",
    "#limits = client.query(query).to_dataframe()\n",
    "limits = run_query(query)\n",
    "\n",
    "print(limits)\n",
    "x = limits.at[0, 'X']\n",
    "y = limits.at[0, 'Y']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After you found the `[X, Y]` range for your gene of interst you can run *Type1* queries efficiently. Here are a couple of examples:\n",
    "\n",
    "### Query 1\n",
    "Find the number of indels and snvs in the region of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2a6579dc563b4ea7a1c9011c8d29b18c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=2.0, style=ProgressStyle(description_wi…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed: 0.00476837158203125¢\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num</th>\n",
       "      <th>variant_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22490</td>\n",
       "      <td>snv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2878</td>\n",
       "      <td>indel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     num variant_type\n",
       "0  22490          snv\n",
       "1   2878        indel"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: For v2_1_1 the \"variant_type\" column must be replaced with \"alternate_bases.allele_type AS variant_type\"\n",
    "query_template = \"\"\"\n",
    "SELECT COUNT(1) AS num, variant_type\n",
    "FROM (\n",
    "SELECT DISTINCT \n",
    "       start_position,\n",
    "       reference_bases,\n",
    "       alternate_bases.alt,\n",
    "       variant_type,\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases\n",
    "WHERE start_position >= {X} AND start_position <= {Y}\n",
    ")\n",
    "GROUP BY 2\n",
    "ORDER BY 1 DESC\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "summary = run_query(query)\n",
    "summary.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of aggregating the results in BigQuery to count the number of each variant type, we could return all rows and process them here. The following query adds a few more columns to the previous query. \n",
    "\n",
    "### Query2\n",
    "A query to retrieve all variants in the region of interest along with `AN` and `AC` values split by gender.\n",
    "\n",
    " * `AN`: Total number of alleles in samples\n",
    " * `AC`: Alternate allele count for samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f86ac20047764b9d94018f01dfc35c75",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=25368.0, style=ProgressStyle(descriptio…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.00476837158203125\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_position</th>\n",
       "      <th>reference_bases</th>\n",
       "      <th>AN</th>\n",
       "      <th>AN_male</th>\n",
       "      <th>AN_female</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>num_alts</th>\n",
       "      <th>AC</th>\n",
       "      <th>AC_male</th>\n",
       "      <th>AC_female</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41191318</td>\n",
       "      <td>T</td>\n",
       "      <td>143350</td>\n",
       "      <td>69482</td>\n",
       "      <td>73868</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41191318</td>\n",
       "      <td>T</td>\n",
       "      <td>143350</td>\n",
       "      <td>69482</td>\n",
       "      <td>73868</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41191327</td>\n",
       "      <td>A</td>\n",
       "      <td>143350</td>\n",
       "      <td>69468</td>\n",
       "      <td>73882</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41191332</td>\n",
       "      <td>G</td>\n",
       "      <td>143310</td>\n",
       "      <td>69444</td>\n",
       "      <td>73866</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>41191336</td>\n",
       "      <td>C</td>\n",
       "      <td>143268</td>\n",
       "      <td>69420</td>\n",
       "      <td>73848</td>\n",
       "      <td>T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start_position reference_bases      AN  AN_male  AN_female alt  \\\n",
       "0        41191318               T  143350    69482      73868   C   \n",
       "1        41191318               T  143350    69482      73868   A   \n",
       "2        41191327               A  143350    69468      73882   C   \n",
       "3        41191332               G  143310    69444      73866   A   \n",
       "4        41191336               C  143268    69420      73848   T   \n",
       "\n",
       "  variant_type  num_alts  AC  AC_male  AC_female  \n",
       "0          snv         1   1        1          0  \n",
       "1          snv         1   3        1          2  \n",
       "2          snv         1   5        5          0  \n",
       "3          snv         1   1        0          1  \n",
       "4          snv         1   2        0          2  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: For v2_1_1 the \"variant_type\" column must be replaced with \"alternate_bases.allele_type AS variant_type\"\n",
    "query_template = \"\"\"\n",
    "SELECT DISTINCT \n",
    "       start_position,\n",
    "       reference_bases,\n",
    "       AN,\n",
    "       AN_male,\n",
    "       AN_female,\n",
    "       alternate_bases.alt AS alt,\n",
    "       variant_type,\n",
    "       alternate_bases.AC AS AC,\n",
    "       alternate_bases.AC_male AS AC_male,\n",
    "       alternate_bases.AC_female AS AC_female,\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases\n",
    "WHERE start_position >= {X} AND start_position <= {Y}\n",
    "ORDER BY 1,2\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "stats_gender = run_query(query)\n",
    "stats_gender.head()\n",
    "\n",
    "# ToDo: add some analysis to the returned dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of splitting `AN` and `AC` values by gender we can analyze ancestry.\n",
    "\n",
    "### Query3\n",
    "A query to retrieve all variants in the region of interest along with `AN` and `AC` values for the following ancestries:\n",
    "* `afr`: African-American/African ancestry\n",
    "* `amr`: Latino ancestry\n",
    "* `eas`: East Asian ancestry\n",
    "* `nfe`: Non-Finnish European ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "64a965cd14e0425cabe9bc9a095ef557",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=25368.0, style=ProgressStyle(descriptio…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.00476837158203125\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_position</th>\n",
       "      <th>reference_bases</th>\n",
       "      <th>AN_afr</th>\n",
       "      <th>AN_amr</th>\n",
       "      <th>AN_eas</th>\n",
       "      <th>AN_nfe</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>AC_afr</th>\n",
       "      <th>AC_amr</th>\n",
       "      <th>AC_eas</th>\n",
       "      <th>AC_nfe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41191318</td>\n",
       "      <td>T</td>\n",
       "      <td>42052</td>\n",
       "      <td>13658</td>\n",
       "      <td>3134</td>\n",
       "      <td>64594</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41191318</td>\n",
       "      <td>T</td>\n",
       "      <td>42052</td>\n",
       "      <td>13658</td>\n",
       "      <td>3134</td>\n",
       "      <td>64594</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41191327</td>\n",
       "      <td>A</td>\n",
       "      <td>42062</td>\n",
       "      <td>13660</td>\n",
       "      <td>3134</td>\n",
       "      <td>64586</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41191332</td>\n",
       "      <td>G</td>\n",
       "      <td>42048</td>\n",
       "      <td>13662</td>\n",
       "      <td>3134</td>\n",
       "      <td>64596</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>41191336</td>\n",
       "      <td>C</td>\n",
       "      <td>42028</td>\n",
       "      <td>13646</td>\n",
       "      <td>3134</td>\n",
       "      <td>64572</td>\n",
       "      <td>T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start_position reference_bases  AN_afr  AN_amr  AN_eas  AN_nfe alt  \\\n",
       "0        41191318               T   42052   13658    3134   64594   C   \n",
       "1        41191318               T   42052   13658    3134   64594   A   \n",
       "2        41191327               A   42062   13660    3134   64586   C   \n",
       "3        41191332               G   42048   13662    3134   64596   A   \n",
       "4        41191336               C   42028   13646    3134   64572   T   \n",
       "\n",
       "  variant_type  AC_afr  AC_amr  AC_eas  AC_nfe  \n",
       "0          snv       0       0       0       1  \n",
       "1          snv       0       0       0       3  \n",
       "2          snv       4       0       0       0  \n",
       "3          snv       0       0       0       1  \n",
       "4          snv       1       0       0       1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: For v2_1_1 the \"variant_type\" column must be replaced with \"alternate_bases.allele_type AS variant_type\"\n",
    "query_template = \"\"\"\n",
    "SELECT DISTINCT \n",
    "       start_position,\n",
    "       reference_bases,\n",
    "       AN_afr,\n",
    "       AN_amr,\n",
    "       AN_eas,\n",
    "       AN_nfe,\n",
    "       alternate_bases.alt AS alt,\n",
    "       variant_type,\n",
    "       alternate_bases.AC_afr AS AC_afr,\n",
    "       alternate_bases.AC_amr AS AC_amr,\n",
    "       alternate_bases.AC_eas AS AC_eas,\n",
    "       alternate_bases.AC_nfe AS AC_nfe,\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases\n",
    "WHERE start_position >= {X} AND start_position <= {Y}\n",
    "ORDER BY 1,2\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "stats_ancestry = run_query(query)\n",
    "stats_ancestry.head()\n",
    "\n",
    "# ToDo: add some analysis to the returned dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query4\n",
    "Given a region of interest, compute the burden of mutation for the gene along with other summary statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ebbfbb120070427997847476d8810f61",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=1.0, style=ProgressStyle(description_wi…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed: 0.0$\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_of_mutation</th>\n",
       "      <th>num_variants</th>\n",
       "      <th>num_alts</th>\n",
       "      <th>sum_AC</th>\n",
       "      <th>quantiles_AC</th>\n",
       "      <th>sum_AN</th>\n",
       "      <th>quantiles_AN</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.969</td>\n",
       "      <td>34254</td>\n",
       "      <td>34254</td>\n",
       "      <td>16394200</td>\n",
       "      <td>[0, 0, 1, 1, 1, 1, 2, 3, 5, 19, 143046]</td>\n",
       "      <td>4625933720</td>\n",
       "      <td>[274, 122910, 140022, 142438, 142940, 143120, ...</td>\n",
       "      <td>BRCA1, NBR2, NBR1, AC060780.1, AC060780.2, AC0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   burden_of_mutation  num_variants  num_alts    sum_AC  \\\n",
       "0               3.969         34254     34254  16394200   \n",
       "\n",
       "                              quantiles_AC      sum_AN  \\\n",
       "0  [0, 0, 1, 1, 1, 1, 2, 3, 5, 19, 143046]  4625933720   \n",
       "\n",
       "                                        quantiles_AN  \\\n",
       "0  [274, 122910, 140022, 142438, 142940, 143120, ...   \n",
       "\n",
       "                                               genes  \n",
       "0  BRCA1, NBR2, NBR1, AC060780.1, AC060780.2, AC0...  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_template = \"\"\"\n",
    "WITH summary_stats AS (\n",
    "SELECT\n",
    "  COUNT(1) AS num_variants,\n",
    "  SUM(ARRAY_LENGTH(alternate_bases)) AS num_alts,  # This data appears to be bi-allelic.\n",
    "  SUM((SELECT alt.AC FROM UNNEST(alternate_bases) AS alt)) AS sum_AC,\n",
    "  APPROX_QUANTILES((SELECT alt.AC FROM UNNEST(alternate_bases) AS alt), 10) AS quantiles_AC,\n",
    "  SUM(AN) AS sum_AN,\n",
    "  APPROX_QUANTILES(AN, 10) AS quantiles_AN,\n",
    "  -- Also include some information from Variant Effect Predictor (VEP).\n",
    "  STRING_AGG(DISTINCT (SELECT annot.symbol FROM UNNEST(alternate_bases) AS alt,\n",
    "                                                UNNEST(vep) AS annot LIMIT 1), ', ') AS genes\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table\n",
    "WHERE start_position >= {X} AND start_position <= {Y})\n",
    "---\n",
    "--- The resulting quantiles and burden_of_mutation score give a very rough idea of the mutation\n",
    "--- rate within these particular regions of the genome. This query could be further refined to\n",
    "--- compute over smaller windows within the regions of interest and/or over different groupings\n",
    "--- of AC and AN by population.\n",
    "---\n",
    "SELECT\n",
    "  ROUND(({Y} - {X}) / num_variants, 3) AS burden_of_mutation,\n",
    "  *,\n",
    "FROM summary_stats\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "stats = run_query(query)\n",
    "stats.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type2: Explore an entire chromosome\n",
    "## Query1\n",
    "Find 10,000 SNV on chr17 that are more common in women than men, min sample size set to 30."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery result1 --use_rest_api\n",
    "SELECT DISTINCT \n",
    "       start_position AS str_pos,\n",
    "       reference_bases AS ref,\n",
    "       alternate_bases.alt AS alt,\n",
    "       alternate_bases.allele_type AS type,\n",
    "       vep.SYMBOL AS gene,\n",
    "       vep.feature_type AS f_type,\n",
    "       alternate_bases.AC AS AC,\n",
    "       alternate_bases.AC_female AS AC_f,\n",
    "       alternate_bases.AC_male AS AC_m,\n",
    "       ROUND(alternate_bases.AC_female / alternate_bases.AC, 3) AS f_ratio\n",
    "FROM `bigquery-public-data.gnomAD.v2_1_1_exomes__chr17` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE alternate_bases.AC > 30 AND vep.SYMBOL IS NOT NULL\n",
    "ORDER BY f_ratio DESC\n",
    "LIMIT 10000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>gene</th>\n",
       "      <th>f_type</th>\n",
       "      <th>AC</th>\n",
       "      <th>AC_f</th>\n",
       "      <th>AC_m</th>\n",
       "      <th>f_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>79140505</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>AATK</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>41</td>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>0.951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>79140505</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>AATK-AS1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>41</td>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>0.951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33998749</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>ins</td>\n",
       "      <td>AP2B1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>37</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7221462</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>snv</td>\n",
       "      <td>GPS2</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>80</td>\n",
       "      <td>75</td>\n",
       "      <td>5</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7221462</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>snv</td>\n",
       "      <td>NEURL4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>80</td>\n",
       "      <td>75</td>\n",
       "      <td>5</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    str_pos ref alt type      gene      f_type  AC  AC_f  AC_m  f_ratio\n",
       "0  79140505   A   C  snv      AATK  Transcript  41    39     2    0.951\n",
       "1  79140505   A   C  snv  AATK-AS1  Transcript  41    39     2    0.951\n",
       "2  33998749   T  TC  ins     AP2B1  Transcript  37    35     2    0.946\n",
       "3   7221462   T   G  snv      GPS2  Transcript  80    75     5    0.938\n",
       "4   7221462   T   G  snv    NEURL4  Transcript  80    75     5    0.938"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can condensed the result and only list gene symbols and the number of variants found in the query1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_pos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNAH17</th>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTC-297N7.11</th>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-799N11.1</th>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF213</th>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNAH9</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_pos\n",
       "gene                  \n",
       "DNAH17              73\n",
       "CTC-297N7.11        67\n",
       "RP11-799N11.1       62\n",
       "RNF213              59\n",
       "DNAH9               54"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result1.groupby('gene').count()[['str_pos']].sort_values(by=['str_pos'], ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query2\n",
    "Find top 1,000 SNV on chr17 that show the most significant differences between male samples of African-American ancestry versus Finnish ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery result2 --use_rest_api\n",
    "SELECT DISTINCT \n",
    "       start_position AS str_pos,\n",
    "       reference_bases AS ref,\n",
    "       alternate_bases.alt AS alt,\n",
    "       alternate_bases.allele_type AS type,\n",
    "       vep.SYMBOL AS gene,\n",
    "       vep.feature_type AS f_type,\n",
    "       alternate_bases.AC_male AS AC_m,\n",
    "       alternate_bases.AC_fin_male AS AC_fin_m,\n",
    "       alternate_bases.AC_afr_male AS AC_afr_m,\n",
    "       ROUND(ABS(alternate_bases.AC_fin_male - alternate_bases.AC_afr_male) / alternate_bases.AC_male, 3) AS fin_afr_diff\n",
    "FROM `bigquery-public-data.gnomAD.v2_1_1_exomes__chr17` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE vep.SYMBOL IS NOT NULL AND\n",
    "      alternate_bases.AC_male > 20 AND alternate_bases.AC_fin_male > 0 AND alternate_bases.AC_afr_male > 0\n",
    "order by fin_afr_diff DESC\n",
    "LIMIT 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>gene</th>\n",
       "      <th>f_type</th>\n",
       "      <th>AC_m</th>\n",
       "      <th>AC_fin_m</th>\n",
       "      <th>AC_afr_m</th>\n",
       "      <th>fin_afr_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>76557038</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>DNAH17</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>56</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>0.929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>76116856</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>snv</td>\n",
       "      <td>TMC6</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>247</td>\n",
       "      <td>228</td>\n",
       "      <td>1</td>\n",
       "      <td>0.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6900268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>ALOX12</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6900268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>RP11-589P10.5</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6900268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>RP11-589P10.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    str_pos ref alt type           gene      f_type  AC_m  AC_fin_m  AC_afr_m  \\\n",
       "0  76557038   G   A  snv         DNAH17  Transcript    56        53         1   \n",
       "1  76116856   C   T  snv           TMC6  Transcript   247       228         1   \n",
       "2   6900268   G   A  snv         ALOX12  Transcript    34        32         1   \n",
       "3   6900268   G   A  snv  RP11-589P10.5  Transcript    34        32         1   \n",
       "4   6900268   G   A  snv  RP11-589P10.7  Transcript    34        32         1   \n",
       "\n",
       "   fin_afr_diff  \n",
       "0         0.929  \n",
       "1         0.919  \n",
       "2         0.912  \n",
       "3         0.912  \n",
       "4         0.912  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query3\n",
    "Find top 1000 genes with the highest number of SNV on chr17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery result3 --use_rest_api\n",
    "SELECT gene, count(1) AS num_snv\n",
    "FROM\n",
    "(\n",
    "SELECT DISTINCT \n",
    "       start_position AS str_pos,\n",
    "       alternate_bases.alt AS alt,\n",
    "       vep.SYMBOL AS gene,\n",
    "FROM `bigquery-public-data.gnomAD.v2_1_1_exomes__chr17` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE vep.SYMBOL IS NOT NULL AND alternate_bases.allele_type = 'snv'\n",
    ")\n",
    "GROUP BY 1\n",
    "ORDER BY 2 DESC\n",
    "LIMIT 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>num_snv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTC-297N7.11</td>\n",
       "      <td>9589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNAH17</td>\n",
       "      <td>9208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RP11-799N11.1</td>\n",
       "      <td>9190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RNF213</td>\n",
       "      <td>6561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNAH2</td>\n",
       "      <td>6361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene  num_snv\n",
       "0   CTC-297N7.11     9589\n",
       "1         DNAH17     9208\n",
       "2  RP11-799N11.1     9190\n",
       "3         RNF213     6561\n",
       "4          DNAH2     6361"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result3.head()"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "name": "common-cpu.m56",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/base-cpu:m56"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
