{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The google.cloud.bigquery extension is already loaded. To reload it, use:\n",
      "  %reload_ext google.cloud.bigquery\n"
     ]
    }
   ],
   "source": [
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gnomAD Queries\n",
    "## Type1: Explore a particular genomic region\n",
    "This category include queries that extract information from a region of genome, for example a gene. Becuase gnomAD BigQuery tables utilize [integer range partitioning](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions) they are optimized for this type of queries.\n",
    "\n",
    "The main requirement to use this feature is to limit queries to a particular region by adding these conditions to the `WHERE` clause:\n",
    "\n",
    "`WHERE start_position >= X AND start_position <= Y`\n",
    "\n",
    "Where `[X, Y]` the region of interst. \n",
    "\n",
    "You can find values of `X` and `Y` by refering to an external databses. For example following table sumarizes the start and end positions for 4 genes on chromosome 17 extracted from an external resource:\n",
    "\n",
    "| Gene \t| X \t| Y \t| Source \t|\n",
    "|:-:\t|-\t|-\t|-\t|\n",
    "| BRCA1 \t| 43044295 \t| 43125364 \t| [link](https://ghr.nlm.nih.gov/gene/BRCA1#location) \t|\n",
    "| COL1A1 \t| 50184096 \t| 50201649 \t| [link](https://ghr.nlm.nih.gov/gene/COL1A1#location) \t|\n",
    "| TP53 \t| 31094927 \t| 31377677 \t| [link](https://ghr.nlm.nih.gov/gene/TP53#location) \t|\n",
    "| NF1 \t| 56593699 \t| 56595611 \t| [link](https://ghr.nlm.nih.gov/gene/NF1#location) \t|\n",
    "\n",
    "Alternatively you could use the following query that extract the same infomration directly from gnomAD tables. \n",
    "\n",
    "In the following example we are using `BRCA1` on `chr17` as example, you could enter your gene of interest to modify all the following queries. Make sure for your gene of interest you are querying the right table (chromosome). If your query returns `NaN` this might be becuase you are querying the wrong table.\n",
    "\n",
    "Also you could choose which version of gnomAD dataset you'd like to user for all the following queries:\n",
    " * `v2_1_1_exomes`\n",
    " * `v2_1_1_genomes`\n",
    " * `v3_genomes`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e961a6191f234c02a178eda9ad4ce419",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dropdown(description='chromosome:', index=16, options=('chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7',…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ToDo: Use ipywidgets to get inputs from user\n",
    "# https://ipywidgets.readthedocs.io/en/latest/examples/Using%20Interact.html\n",
    "\n",
    "import ipywidgets as widgets\n",
    "\n",
    "widgets.Dropdown(\n",
    "    options=['v2_1_1_exomes', 'v2_1_1_genomes', 'v3_genomes'],\n",
    "    value='v3_genomes',\n",
    "    description='gnomAD version:',\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "widgets.Dropdown(\n",
    "    options=['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY'],\n",
    "    value='chr17',\n",
    "    description='chromosome:',\n",
    "    disabled=False,\n",
    ")\n",
    "\n",
    "widgets.Text(\n",
    "    value='BRCA1',\n",
    "    placeholder='gene_symbol',\n",
    "    description='Gene Symbol:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "from ipywidgets import interact\n",
    "\n",
    "def f(x):\n",
    "    return x\n",
    "\n",
    "interact(f, x='BRCA1');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ToDo: Use ipywidgets to get inputs from user\n",
    "chromosome = 'chr4'\n",
    "gene_symbol = 'SNCA'#'BRCA1'\n",
    "gnomad_version = 'v3_genomes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from google.cloud import bigquery\n",
    "\n",
    "client = bigquery.Client()\n",
    "\n",
    "def run_query(query):\n",
    "    query_job = client.query(query)\n",
    "    result = query_job.to_dataframe(progress_bar_type='tqdm_notebook')\n",
    "    cost_cents = (query_job.total_bytes_billed / 1024 ** 4) * 500\n",
    "    print('This query costed (¢): {}'.format(cost_cents))\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8245202b3ff54034a350b88f77e1a58d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=1.0, style=ProgressStyle(description_wi…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.8034706115722656\n",
      "          X         Y\n",
      "0  89719100  89843314\n"
     ]
    }
   ],
   "source": [
    "query_template = \"\"\"\n",
    "SELECT MIN(start_position) AS X, MAX(end_position) AS Y\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table\n",
    "WHERE EXISTS\n",
    "  (SELECT 1 FROM UNNEST(main_table.alternate_bases) AS alternate_bases\n",
    "   WHERE EXISTS (SELECT 1 from alternate_bases.vep WHERE SYMBOL = '{GENE}'))\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, GENE=gene_symbol)\n",
    "\n",
    "#limits = client.query(query).to_dataframe()\n",
    "limits = run_query(query)\n",
    "\n",
    "print(limits)\n",
    "x = limits.at[0, 'X']\n",
    "y = limits.at[0, 'Y']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After you found the `[X, Y]` range for your gene of interst you can run *Type1* queries efficiently. Here are a couple of examples:\n",
    "\n",
    "### Query 1\n",
    "Find the number of indels and snvs in the region of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2a6579dc563b4ea7a1c9011c8d29b18c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=2.0, style=ProgressStyle(description_wi…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed: 0.00476837158203125¢\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num</th>\n",
       "      <th>variant_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22490</td>\n",
       "      <td>snv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2878</td>\n",
       "      <td>indel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     num variant_type\n",
       "0  22490          snv\n",
       "1   2878        indel"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: For v2_1_1 the \"variant_type\" column must be replaced with \"alternate_bases.allele_type AS variant_type\"\n",
    "query_template = \"\"\"\n",
    "SELECT COUNT(1) AS num, variant_type\n",
    "FROM (\n",
    "SELECT DISTINCT \n",
    "       start_position,\n",
    "       reference_bases,\n",
    "       alternate_bases.alt,\n",
    "       variant_type,\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases\n",
    "WHERE start_position >= {X} AND start_position <= {Y}\n",
    ")\n",
    "GROUP BY 2\n",
    "ORDER BY 1 DESC\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "summary = run_query(query)\n",
    "summary.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of aggregating the results in BigQuery to count the number of each variant type, we could return all rows and process them here. The following query adds a few more columns to the previous query. \n",
    "\n",
    "### Query2\n",
    "A query to retrieve all variants in the region of interest along with `AN` and `AC` values split by gender.\n",
    "\n",
    " * `AN`: Total number of alleles in samples\n",
    " * `AC`: Alternate allele count for samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "45670dea3584439a957e04c9b18a8272",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=25368.0, style=ProgressStyle(descriptio…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.00476837158203125\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>reference_bases</th>\n",
       "      <th>names</th>\n",
       "      <th>AN</th>\n",
       "      <th>AN_male</th>\n",
       "      <th>AN_female</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>AC</th>\n",
       "      <th>AC_male</th>\n",
       "      <th>AC_female</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>41191318</td>\n",
       "      <td>41191319</td>\n",
       "      <td>T</td>\n",
       "      <td>[]</td>\n",
       "      <td>143350</td>\n",
       "      <td>69482</td>\n",
       "      <td>73868</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>41191318</td>\n",
       "      <td>41191319</td>\n",
       "      <td>T</td>\n",
       "      <td>[rs1037204244]</td>\n",
       "      <td>143350</td>\n",
       "      <td>69482</td>\n",
       "      <td>73868</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>41191327</td>\n",
       "      <td>41191328</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>143350</td>\n",
       "      <td>69468</td>\n",
       "      <td>73882</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>41191332</td>\n",
       "      <td>41191333</td>\n",
       "      <td>G</td>\n",
       "      <td>[rs1389611149]</td>\n",
       "      <td>143310</td>\n",
       "      <td>69444</td>\n",
       "      <td>73866</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>41191336</td>\n",
       "      <td>41191337</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs1404669652]</td>\n",
       "      <td>143268</td>\n",
       "      <td>69420</td>\n",
       "      <td>73848</td>\n",
       "      <td>T</td>\n",
       "      <td>snv</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reference_name  start_position  end_position reference_bases  \\\n",
       "0          chr17        41191318      41191319               T   \n",
       "1          chr17        41191318      41191319               T   \n",
       "2          chr17        41191327      41191328               A   \n",
       "3          chr17        41191332      41191333               G   \n",
       "4          chr17        41191336      41191337               C   \n",
       "\n",
       "            names      AN  AN_male  AN_female alt variant_type  AC  AC_male  \\\n",
       "0              []  143350    69482      73868   C          snv   1        1   \n",
       "1  [rs1037204244]  143350    69482      73868   A          snv   3        1   \n",
       "2              []  143350    69468      73882   C          snv   5        5   \n",
       "3  [rs1389611149]  143310    69444      73866   A          snv   1        0   \n",
       "4  [rs1404669652]  143268    69420      73848   T          snv   2        0   \n",
       "\n",
       "   AC_female  \n",
       "0          0  \n",
       "1          2  \n",
       "2          0  \n",
       "3          1  \n",
       "4          2  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: For v2_1_1 the \"variant_type\" column must be replaced with \"alternate_bases.allele_type AS variant_type\"\n",
    "query_template = \"\"\"\n",
    "SELECT reference_name, \n",
    "       start_position,\n",
    "       end_position,\n",
    "       reference_bases,\n",
    "       names,\n",
    "       AN,\n",
    "       AN_male,\n",
    "       AN_female,\n",
    "       alternate_bases.alt AS alt,\n",
    "       variant_type,\n",
    "       alternate_bases.AC AS AC,\n",
    "       alternate_bases.AC_male AS AC_male,\n",
    "       alternate_bases.AC_female AS AC_female,\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases\n",
    "WHERE start_position >= {X} AND start_position <= {Y}\n",
    "ORDER BY start_position, reference_bases\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "stats_gender = run_query(query)\n",
    "stats_gender.head()\n",
    "\n",
    "# ToDo: add some analysis to the returned dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of splitting `AN` and `AC` values by gender we can analyze ancestry.\n",
    "\n",
    "### Query3\n",
    "A query to retrieve all variants in the region of interest along with `AN` and `AC` values for the following ancestries:\n",
    "* `afr`: African-American/African ancestry\n",
    "* `amr`: Latino ancestry\n",
    "* `eas`: East Asian ancestry\n",
    "* `nfe`: Non-Finnish European ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "df6cf4a0da9e471b8d7da816ee061cf2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=25368.0, style=ProgressStyle(descriptio…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.00476837158203125\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>reference_bases</th>\n",
       "      <th>names</th>\n",
       "      <th>AN_afr</th>\n",
       "      <th>AN_amr</th>\n",
       "      <th>AN_eas</th>\n",
       "      <th>AN_nfe</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>AC_afr</th>\n",
       "      <th>AC_amr</th>\n",
       "      <th>AC_eas</th>\n",
       "      <th>AC_nfe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41191318</td>\n",
       "      <td>41191319</td>\n",
       "      <td>T</td>\n",
       "      <td>[]</td>\n",
       "      <td>42052</td>\n",
       "      <td>13658</td>\n",
       "      <td>3134</td>\n",
       "      <td>64594</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41191318</td>\n",
       "      <td>41191319</td>\n",
       "      <td>T</td>\n",
       "      <td>[rs1037204244]</td>\n",
       "      <td>42052</td>\n",
       "      <td>13658</td>\n",
       "      <td>3134</td>\n",
       "      <td>64594</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41191327</td>\n",
       "      <td>41191328</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>42062</td>\n",
       "      <td>13660</td>\n",
       "      <td>3134</td>\n",
       "      <td>64586</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41191332</td>\n",
       "      <td>41191333</td>\n",
       "      <td>G</td>\n",
       "      <td>[rs1389611149]</td>\n",
       "      <td>42048</td>\n",
       "      <td>13662</td>\n",
       "      <td>3134</td>\n",
       "      <td>64596</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>41191336</td>\n",
       "      <td>41191337</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs1404669652]</td>\n",
       "      <td>42028</td>\n",
       "      <td>13646</td>\n",
       "      <td>3134</td>\n",
       "      <td>64572</td>\n",
       "      <td>T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start_position  end_position reference_bases           names  AN_afr  \\\n",
       "0        41191318      41191319               T              []   42052   \n",
       "1        41191318      41191319               T  [rs1037204244]   42052   \n",
       "2        41191327      41191328               A              []   42062   \n",
       "3        41191332      41191333               G  [rs1389611149]   42048   \n",
       "4        41191336      41191337               C  [rs1404669652]   42028   \n",
       "\n",
       "   AN_amr  AN_eas  AN_nfe alt variant_type  AC_afr  AC_amr  AC_eas  AC_nfe  \n",
       "0   13658    3134   64594   C          snv       0       0       0       1  \n",
       "1   13658    3134   64594   A          snv       0       0       0       3  \n",
       "2   13660    3134   64586   C          snv       4       0       0       0  \n",
       "3   13662    3134   64596   A          snv       0       0       0       1  \n",
       "4   13646    3134   64572   T          snv       1       0       0       1  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOTE: For v2_1_1 the \"variant_type\" column must be replaced with \"alternate_bases.allele_type AS variant_type\"\n",
    "query_template = \"\"\"\n",
    "SELECT reference_name, \n",
    "       start_position,\n",
    "       end_position,\n",
    "       reference_bases,\n",
    "       names,\n",
    "       AN_afr,\n",
    "       AN_amr,\n",
    "       AN_eas,\n",
    "       AN_nfe,\n",
    "       alternate_bases.alt AS alt,\n",
    "       variant_type,\n",
    "       alternate_bases.AC_afr AS AC_afr,\n",
    "       alternate_bases.AC_amr AS AC_amr,\n",
    "       alternate_bases.AC_eas AS AC_eas,\n",
    "       alternate_bases.AC_nfe AS AC_nfe,\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases\n",
    "WHERE start_position >= {X} AND start_position <= {Y}\n",
    "ORDER BY start_position, reference_bases\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "stats_ancestry = run_query(query)\n",
    "stats_ancestry.head()\n",
    "\n",
    "# ToDo: add some analysis to the returned dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query4\n",
    "gnomAD tables have many more columns, you can find the full list of columns along with their description using the following query.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "af4a5addd3bb4375a0caf8485c21efaf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=649.0, style=ProgressStyle(description_…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.0095367431640625\n",
      "There are 649 columns in `bigquery-public-data.gnomAD.v3_genomes__chr17` table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>column_name</th>\n",
       "      <th>field_path</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>reference_name</td>\n",
       "      <td>reference_name</td>\n",
       "      <td>Reference name.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>start_position</td>\n",
       "      <td>start_position</td>\n",
       "      <td>Start position (0-based). Corresponds to the f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>end_position</td>\n",
       "      <td>end_position</td>\n",
       "      <td>End position (0-based). Corresponds to the fir...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>reference_bases</td>\n",
       "      <td>reference_bases</td>\n",
       "      <td>Reference bases.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>alternate_bases</td>\n",
       "      <td>alternate_bases</td>\n",
       "      <td>One record for each alternate base (if any).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>alternate_bases</td>\n",
       "      <td>alternate_bases.alt</td>\n",
       "      <td>Alternate base.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>alternate_bases</td>\n",
       "      <td>alternate_bases.AC</td>\n",
       "      <td>Alternate allele count for samples</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       column_name           field_path  \\\n",
       "0   reference_name       reference_name   \n",
       "1   start_position       start_position   \n",
       "2     end_position         end_position   \n",
       "3  reference_bases      reference_bases   \n",
       "4  alternate_bases      alternate_bases   \n",
       "5  alternate_bases  alternate_bases.alt   \n",
       "6  alternate_bases   alternate_bases.AC   \n",
       "\n",
       "                                         description  \n",
       "0                                    Reference name.  \n",
       "1  Start position (0-based). Corresponds to the f...  \n",
       "2  End position (0-based). Corresponds to the fir...  \n",
       "3                                   Reference bases.  \n",
       "4       One record for each alternate base (if any).  \n",
       "5                                    Alternate base.  \n",
       "6                 Alternate allele count for samples  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_template = \"\"\"\n",
    "SELECT column_name, field_path, description\n",
    "FROM `bigquery-public-data`.gnomAD.INFORMATION_SCHEMA.COLUMN_FIELD_PATHS\n",
    "WHERE table_name = \"v2_1_1_genomes__chr17\"\n",
    "      AND column_name IN (\n",
    "          SELECT COLUMN_NAME\n",
    "          FROM `bigquery-public-data`.gnomAD.INFORMATION_SCHEMA.COLUMNS\n",
    "          WHERE table_name = \"v2_1_1_genomes__chr17\")\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "column_info = run_query(query)\n",
    "print('There are {} columns in `bigquery-public-data.gnomAD.{}__{}` table'.format(len(column_info.index), gnomad_version, chromosome))\n",
    "column_info.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using `column_info` dataframe you can find other available values for the ancestry slice:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>column_name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>AN</td>\n",
       "      <td>Total number of alleles in samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>AN_afr</td>\n",
       "      <td>Total number of alleles in samples of African-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>AN_afr_female</td>\n",
       "      <td>Total number of alleles in female samples of A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>AN_afr_male</td>\n",
       "      <td>Total number of alleles in male samples of Afr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>AN_amr</td>\n",
       "      <td>Total number of alleles in samples of Latino a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>AN_amr_female</td>\n",
       "      <td>Total number of alleles in female samples of L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>AN_amr_male</td>\n",
       "      <td>Total number of alleles in male samples of Lat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>AN_asj</td>\n",
       "      <td>Total number of alleles in samples of Ashkenaz...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>AN_asj_female</td>\n",
       "      <td>Total number of alleles in female samples of A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>AN_asj_male</td>\n",
       "      <td>Total number of alleles in male samples of Ash...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>AN_eas</td>\n",
       "      <td>Total number of alleles in samples of East Asi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       column_name                                        description\n",
       "506             AN                 Total number of alleles in samples\n",
       "542         AN_afr  Total number of alleles in samples of African-...\n",
       "547  AN_afr_female  Total number of alleles in female samples of A...\n",
       "541    AN_afr_male  Total number of alleles in male samples of Afr...\n",
       "555         AN_amr  Total number of alleles in samples of Latino a...\n",
       "642  AN_amr_female  Total number of alleles in female samples of L...\n",
       "641    AN_amr_male  Total number of alleles in male samples of Lat...\n",
       "625         AN_asj  Total number of alleles in samples of Ashkenaz...\n",
       "622  AN_asj_female  Total number of alleles in female samples of A...\n",
       "566    AN_asj_male  Total number of alleles in male samples of Ash...\n",
       "557         AN_eas  Total number of alleles in samples of East Asi..."
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AN_columns = column_info[column_info['column_name'].str.startswith('AN')] # Retain only rows that column_name starts with \"AN\"\n",
    "AN_columns = AN_columns[['column_name', 'description']] # Drop extra column (field_path)\n",
    "AN_columns = AN_columns.sort_values(by=['column_name']) # Sort by column_name\n",
    "AN_columns.head(11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the corresponding values for `AC` and `AF` (Alternate allele frequency) exist under the `alternate_bases` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>field_path</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>alternate_bases.AC</td>\n",
       "      <td>Alternate allele count for samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>alternate_bases.AC_afr</td>\n",
       "      <td>Alternate allele count for samples of African-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>alternate_bases.AC_afr_female</td>\n",
       "      <td>Alternate allele count for female samples of A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>alternate_bases.AC_afr_male</td>\n",
       "      <td>Alternate allele count for male samples of Afr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>alternate_bases.AC_amr</td>\n",
       "      <td>Alternate allele count for samples of Latino a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>alternate_bases.AC_amr_female</td>\n",
       "      <td>Alternate allele count for female samples of L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>alternate_bases.AC_amr_male</td>\n",
       "      <td>Alternate allele count for male samples of Lat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>alternate_bases.AC_asj</td>\n",
       "      <td>Alternate allele count for samples of Ashkenaz...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>alternate_bases.AC_asj_female</td>\n",
       "      <td>Alternate allele count for female samples of A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>alternate_bases.AC_asj_male</td>\n",
       "      <td>Alternate allele count for male samples of Ash...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>alternate_bases.AC_eas</td>\n",
       "      <td>Alternate allele count for samples of East Asi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        field_path  \\\n",
       "6               alternate_bases.AC   \n",
       "42          alternate_bases.AC_afr   \n",
       "57   alternate_bases.AC_afr_female   \n",
       "39     alternate_bases.AC_afr_male   \n",
       "81          alternate_bases.AC_amr   \n",
       "343  alternate_bases.AC_amr_female   \n",
       "340    alternate_bases.AC_amr_male   \n",
       "292         alternate_bases.AC_asj   \n",
       "283  alternate_bases.AC_asj_female   \n",
       "115    alternate_bases.AC_asj_male   \n",
       "87          alternate_bases.AC_eas   \n",
       "\n",
       "                                           description  \n",
       "6                   Alternate allele count for samples  \n",
       "42   Alternate allele count for samples of African-...  \n",
       "57   Alternate allele count for female samples of A...  \n",
       "39   Alternate allele count for male samples of Afr...  \n",
       "81   Alternate allele count for samples of Latino a...  \n",
       "343  Alternate allele count for female samples of L...  \n",
       "340  Alternate allele count for male samples of Lat...  \n",
       "292  Alternate allele count for samples of Ashkenaz...  \n",
       "283  Alternate allele count for female samples of A...  \n",
       "115  Alternate allele count for male samples of Ash...  \n",
       "87   Alternate allele count for samples of East Asi...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AC_columns = column_info[column_info['field_path'].str.startswith('alternate_bases.AC')] # Retain only rows that field_path starts with \"alternate_bases.AC\"\n",
    "AC_columns = AC_columns[['field_path', 'description']] # Drop extra column (column_name)\n",
    "AC_columns = AC_columns.sort_values(by=['field_path']) # Sort by field_path\n",
    "AC_columns.head(11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next query showcases how to use `AN` and `AC` values.\n",
    "\n",
    "### Query5\n",
    "Given a region of interest, compute the burden of mutation for the gene along with other summary statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "72b46d87302c44a5b31e26611a2b5520",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=1.0, style=ProgressStyle(description_wi…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.00476837158203125\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>burden_of_mutation</th>\n",
       "      <th>num_variants</th>\n",
       "      <th>num_alts</th>\n",
       "      <th>sum_AC</th>\n",
       "      <th>quantiles_AC</th>\n",
       "      <th>sum_AN</th>\n",
       "      <th>quantiles_AN</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.594</td>\n",
       "      <td>25368</td>\n",
       "      <td>25368</td>\n",
       "      <td>22578697</td>\n",
       "      <td>[0, 0, 1, 1, 1, 1, 2, 2, 5, 20, 143221]</td>\n",
       "      <td>3211329782</td>\n",
       "      <td>[174, 70788, 122774, 139282, 142122, 143092, 1...</td>\n",
       "      <td>KRTAP9-12P, KRTAP9-2, KRTAP9-7, KRTAP9-6, KRTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   burden_of_mutation  num_variants  num_alts    sum_AC  \\\n",
       "0               3.594         25368     25368  22578697   \n",
       "\n",
       "                              quantiles_AC      sum_AN  \\\n",
       "0  [0, 0, 1, 1, 1, 1, 2, 2, 5, 20, 143221]  3211329782   \n",
       "\n",
       "                                        quantiles_AN  \\\n",
       "0  [174, 70788, 122774, 139282, 142122, 143092, 1...   \n",
       "\n",
       "                                               genes  \n",
       "0  KRTAP9-12P, KRTAP9-2, KRTAP9-7, KRTAP9-6, KRTA...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_template = \"\"\"\n",
    "WITH summary_stats AS (\n",
    "SELECT\n",
    "  COUNT(1) AS num_variants,\n",
    "  SUM(ARRAY_LENGTH(alternate_bases)) AS num_alts,  # This data appears to be bi-allelic.\n",
    "  SUM((SELECT alt.AC FROM UNNEST(alternate_bases) AS alt)) AS sum_AC,\n",
    "  APPROX_QUANTILES((SELECT alt.AC FROM UNNEST(alternate_bases) AS alt), 10) AS quantiles_AC,\n",
    "  SUM(AN) AS sum_AN,\n",
    "  APPROX_QUANTILES(AN, 10) AS quantiles_AN,\n",
    "  -- Also include some information from Variant Effect Predictor (VEP).\n",
    "  STRING_AGG(DISTINCT (SELECT annot.symbol FROM UNNEST(alternate_bases) AS alt,\n",
    "                                                UNNEST(vep) AS annot LIMIT 1), ', ') AS genes\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table\n",
    "WHERE start_position >= {X} AND start_position <= {Y})\n",
    "---\n",
    "--- The resulting quantiles and burden_of_mutation score give a very rough idea of the mutation\n",
    "--- rate within these particular regions of the genome. This query could be further refined to\n",
    "--- compute over smaller windows within the regions of interest and/or over different groupings\n",
    "--- of AC and AN by population.\n",
    "---\n",
    "SELECT\n",
    "  ROUND(({Y} - {X}) / num_variants, 3) AS burden_of_mutation,\n",
    "  *,\n",
    "FROM summary_stats\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "burden_of_mu = run_query(query)\n",
    "burden_of_mu.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The other column to use is `alternate_bases.vep` which contains the [VEP annotaions](https://uswest.ensembl.org/info/docs/tools/vep/index.html) for each variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>field_path</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>alternate_bases.vep</td>\n",
       "      <td>List of vep annotations for this alternate.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>alternate_bases.vep.allele</td>\n",
       "      <td>The ALT part of the annotation field.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>alternate_bases.vep.Consequence</td>\n",
       "      <td>Consequence type of this variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>alternate_bases.vep.IMPACT</td>\n",
       "      <td>The impact modifier for the consequence type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>alternate_bases.vep.SYMBOL</td>\n",
       "      <td>The gene symbol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          field_path  \\\n",
       "430              alternate_bases.vep   \n",
       "431       alternate_bases.vep.allele   \n",
       "432  alternate_bases.vep.Consequence   \n",
       "433       alternate_bases.vep.IMPACT   \n",
       "434       alternate_bases.vep.SYMBOL   \n",
       "\n",
       "                                      description  \n",
       "430   List of vep annotations for this alternate.  \n",
       "431         The ALT part of the annotation field.  \n",
       "432              Consequence type of this variant  \n",
       "433  The impact modifier for the consequence type  \n",
       "434                               The gene symbol  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_columns = column_info[column_info['field_path'].str.startswith('alternate_bases.vep')] # Retain only rows that field_path starts with \"alternate_bases.vep\"\n",
    "vep_columns = vep_columns[['field_path', 'description']] # Drop extra column (column_name)\n",
    "vep_columns.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next query showcases how to use some of the `vep` annotation values.\n",
    "\n",
    "### Query6\n",
    "Given a region of interest, examine `vep` annotations to pull out variants with negative consequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "69e9e75e5c244cb2b10fed336452c350",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Downloading', max=327.0, style=ProgressStyle(description_…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This query costed (¢): 0.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>reference_bases</th>\n",
       "      <th>names</th>\n",
       "      <th>alt</th>\n",
       "      <th>vep_allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89726656</td>\n",
       "      <td>89726657</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs749526106]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89726656</td>\n",
       "      <td>89726657</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs749526106]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89726656</td>\n",
       "      <td>89726657</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs749526106]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89726656</td>\n",
       "      <td>89726657</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs749526106]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89726656</td>\n",
       "      <td>89726657</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs749526106]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89822348</td>\n",
       "      <td>89822349</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs752981956]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89822348</td>\n",
       "      <td>89822349</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs752981956]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89822348</td>\n",
       "      <td>89822349</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs752981956]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89822348</td>\n",
       "      <td>89822349</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs752981956]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89822348</td>\n",
       "      <td>89822349</td>\n",
       "      <td>C</td>\n",
       "      <td>[rs752981956]</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>ENSG00000145335</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference_name  start_position  end_position reference_bases  \\\n",
       "0             chr4        89726656      89726657               C   \n",
       "1             chr4        89726656      89726657               C   \n",
       "2             chr4        89726656      89726657               C   \n",
       "3             chr4        89726656      89726657               C   \n",
       "4             chr4        89726656      89726657               C   \n",
       "..             ...             ...           ...             ...   \n",
       "195           chr4        89822348      89822349               C   \n",
       "196           chr4        89822348      89822349               C   \n",
       "197           chr4        89822348      89822349               C   \n",
       "198           chr4        89822348      89822349               C   \n",
       "199           chr4        89822348      89822349               C   \n",
       "\n",
       "             names alt vep_allele       Consequence    Impact Symbol  \\\n",
       "0    [rs749526106]   T          T  missense_variant  MODERATE   SNCA   \n",
       "1    [rs749526106]   T          T  missense_variant  MODERATE   SNCA   \n",
       "2    [rs749526106]   T          T  missense_variant  MODERATE   SNCA   \n",
       "3    [rs749526106]   T          T  missense_variant  MODERATE   SNCA   \n",
       "4    [rs749526106]   T          T  missense_variant  MODERATE   SNCA   \n",
       "..             ...  ..        ...               ...       ...    ...   \n",
       "195  [rs752981956]   T          T  missense_variant  MODERATE   SNCA   \n",
       "196  [rs752981956]   T          T  missense_variant  MODERATE   SNCA   \n",
       "197  [rs752981956]   T          T  missense_variant  MODERATE   SNCA   \n",
       "198  [rs752981956]   T          T  missense_variant  MODERATE   SNCA   \n",
       "199  [rs752981956]   T          T  missense_variant  MODERATE   SNCA   \n",
       "\n",
       "                Gene  \n",
       "0    ENSG00000145335  \n",
       "1    ENSG00000145335  \n",
       "2    ENSG00000145335  \n",
       "3    ENSG00000145335  \n",
       "4    ENSG00000145335  \n",
       "..               ...  \n",
       "195  ENSG00000145335  \n",
       "196  ENSG00000145335  \n",
       "197  ENSG00000145335  \n",
       "198  ENSG00000145335  \n",
       "199  ENSG00000145335  \n",
       "\n",
       "[200 rows x 11 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_template = \"\"\"\n",
    "SELECT reference_name,\n",
    "       start_position,\n",
    "       end_position,\n",
    "       reference_bases,\n",
    "       names,\n",
    "       alternate_bases.alt AS alt,\n",
    "       vep.Consequence AS Consequence,\n",
    "       vep.IMPACT AS Impact,\n",
    "       vep.SYMBOL AS Symbol,\n",
    "       vep.Gene AS Gene\n",
    "FROM `bigquery-public-data.gnomAD.{GNOMAD_VER}__{CHROM}` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE start_position >= {X} AND start_position <= {Y} AND\n",
    "      REGEXP_CONTAINS(vep.Consequence, r\"missense_variant\")\n",
    "ORDER BY start_position, reference_bases\n",
    "\"\"\"\n",
    "query = query_template.format(GNOMAD_VER=gnomad_version, CHROM=chromosome, X=x, Y=y)\n",
    "\n",
    "neg_variants = run_query(query)\n",
    "neg_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type2: Explore an entire chromosome\n",
    "## Query1\n",
    "Find 10,000 SNV on chr17 that are more common in women than men, min sample size set to 30."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery result1 --use_rest_api\n",
    "SELECT DISTINCT \n",
    "       start_position AS str_pos,\n",
    "       reference_bases AS ref,\n",
    "       alternate_bases.alt AS alt,\n",
    "       alternate_bases.allele_type AS type,\n",
    "       vep.SYMBOL AS gene,\n",
    "       vep.feature_type AS f_type,\n",
    "       alternate_bases.AC AS AC,\n",
    "       alternate_bases.AC_female AS AC_f,\n",
    "       alternate_bases.AC_male AS AC_m,\n",
    "       ROUND(alternate_bases.AC_female / alternate_bases.AC, 3) AS f_ratio\n",
    "FROM `bigquery-public-data.gnomAD.v2_1_1_exomes__chr17` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE alternate_bases.AC > 30 AND vep.SYMBOL IS NOT NULL\n",
    "ORDER BY f_ratio DESC\n",
    "LIMIT 10000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>gene</th>\n",
       "      <th>f_type</th>\n",
       "      <th>AC</th>\n",
       "      <th>AC_f</th>\n",
       "      <th>AC_m</th>\n",
       "      <th>f_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>79140505</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>AATK</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>41</td>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>0.951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>79140505</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>snv</td>\n",
       "      <td>AATK-AS1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>41</td>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>0.951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33998749</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>ins</td>\n",
       "      <td>AP2B1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>37</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0.946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7221462</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>snv</td>\n",
       "      <td>GPS2</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>80</td>\n",
       "      <td>75</td>\n",
       "      <td>5</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7221462</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>snv</td>\n",
       "      <td>NEURL4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>80</td>\n",
       "      <td>75</td>\n",
       "      <td>5</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    str_pos ref alt type      gene      f_type  AC  AC_f  AC_m  f_ratio\n",
       "0  79140505   A   C  snv      AATK  Transcript  41    39     2    0.951\n",
       "1  79140505   A   C  snv  AATK-AS1  Transcript  41    39     2    0.951\n",
       "2  33998749   T  TC  ins     AP2B1  Transcript  37    35     2    0.946\n",
       "3   7221462   T   G  snv      GPS2  Transcript  80    75     5    0.938\n",
       "4   7221462   T   G  snv    NEURL4  Transcript  80    75     5    0.938"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can condensed the result and only list gene symbols and the number of variants found in the query1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_pos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNAH17</th>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTC-297N7.11</th>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-799N11.1</th>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF213</th>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNAH9</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_pos\n",
       "gene                  \n",
       "DNAH17              73\n",
       "CTC-297N7.11        67\n",
       "RP11-799N11.1       62\n",
       "RNF213              59\n",
       "DNAH9               54"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result1.groupby('gene').count()[['str_pos']].sort_values(by=['str_pos'], ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query2\n",
    "Find top 1,000 SNV on chr17 that show the most significant differences between male samples of African-American ancestry versus Finnish ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery result2 --use_rest_api\n",
    "SELECT DISTINCT \n",
    "       start_position AS str_pos,\n",
    "       reference_bases AS ref,\n",
    "       alternate_bases.alt AS alt,\n",
    "       alternate_bases.allele_type AS type,\n",
    "       vep.SYMBOL AS gene,\n",
    "       vep.feature_type AS f_type,\n",
    "       alternate_bases.AC_male AS AC_m,\n",
    "       alternate_bases.AC_fin_male AS AC_fin_m,\n",
    "       alternate_bases.AC_afr_male AS AC_afr_m,\n",
    "       ROUND(ABS(alternate_bases.AC_fin_male - alternate_bases.AC_afr_male) / alternate_bases.AC_male, 3) AS fin_afr_diff\n",
    "FROM `bigquery-public-data.gnomAD.v2_1_1_exomes__chr17` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE vep.SYMBOL IS NOT NULL AND\n",
    "      alternate_bases.AC_male > 20 AND alternate_bases.AC_fin_male > 0 AND alternate_bases.AC_afr_male > 0\n",
    "order by fin_afr_diff DESC\n",
    "LIMIT 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>gene</th>\n",
       "      <th>f_type</th>\n",
       "      <th>AC_m</th>\n",
       "      <th>AC_fin_m</th>\n",
       "      <th>AC_afr_m</th>\n",
       "      <th>fin_afr_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>76557038</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>DNAH17</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>56</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>0.929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>76116856</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>snv</td>\n",
       "      <td>TMC6</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>247</td>\n",
       "      <td>228</td>\n",
       "      <td>1</td>\n",
       "      <td>0.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6900268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>ALOX12</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6900268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>RP11-589P10.5</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6900268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>snv</td>\n",
       "      <td>RP11-589P10.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    str_pos ref alt type           gene      f_type  AC_m  AC_fin_m  AC_afr_m  \\\n",
       "0  76557038   G   A  snv         DNAH17  Transcript    56        53         1   \n",
       "1  76116856   C   T  snv           TMC6  Transcript   247       228         1   \n",
       "2   6900268   G   A  snv         ALOX12  Transcript    34        32         1   \n",
       "3   6900268   G   A  snv  RP11-589P10.5  Transcript    34        32         1   \n",
       "4   6900268   G   A  snv  RP11-589P10.7  Transcript    34        32         1   \n",
       "\n",
       "   fin_afr_diff  \n",
       "0         0.929  \n",
       "1         0.919  \n",
       "2         0.912  \n",
       "3         0.912  \n",
       "4         0.912  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query3\n",
    "Find top 1000 genes with the highest number of SNV on chr17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery result3 --use_rest_api\n",
    "SELECT gene, count(1) AS num_snv\n",
    "FROM\n",
    "(\n",
    "SELECT DISTINCT \n",
    "       start_position AS str_pos,\n",
    "       alternate_bases.alt AS alt,\n",
    "       vep.SYMBOL AS gene,\n",
    "FROM `bigquery-public-data.gnomAD.v2_1_1_exomes__chr17` AS main_table,\n",
    "     main_table.alternate_bases AS alternate_bases,\n",
    "     alternate_bases.vep AS vep\n",
    "WHERE vep.SYMBOL IS NOT NULL AND alternate_bases.allele_type = 'snv'\n",
    ")\n",
    "GROUP BY 1\n",
    "ORDER BY 2 DESC\n",
    "LIMIT 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>num_snv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTC-297N7.11</td>\n",
       "      <td>9589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNAH17</td>\n",
       "      <td>9208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RP11-799N11.1</td>\n",
       "      <td>9190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RNF213</td>\n",
       "      <td>6561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNAH2</td>\n",
       "      <td>6361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene  num_snv\n",
       "0   CTC-297N7.11     9589\n",
       "1         DNAH17     9208\n",
       "2  RP11-799N11.1     9190\n",
       "3         RNF213     6561\n",
       "4          DNAH2     6361"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result3.head()"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "name": "common-cpu.m56",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/base-cpu:m56"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
